app/src/main/java/at/roteskreuz/stopcorona/model/repositories/InfectionMessengerRepository.kt

```diff
@@ -159,39 +160,48 @@ class InfectionMessengerRepositoryImpl(
     }
 
     private suspend fun processKeysBasedOnBatchToken(token: String) {
-        val warningType = WarningType.RED
+        val warningType = quarantineRepository.getCurrentWarningType()
 
         val summary = exposureNotificationRepository.determineRiskWithoutInformingUser(token)
         val configuration =
             configurationRepository.getConfiguration()
                 ?: throw IllegalStateException("we have no configuration values here, it doesn´t make sense to continue")
 
-        when(warningType) {
-
+        when (warningType) {
             WarningType.YELLOW, WarningType.RED -> {
                 //TODO finish this decision branch
-                if (summary.summationRiskScore >= configuration.dailyRiskThreshold){
+                if (summary.summationRiskScore >= configuration.dailyRiskThreshold) {
                     val summary = exposureNotificationRepository.getExposureSummaryWithPotentiallyInformingTheUser(token)
                     //go through the days and check if the day is the first RED/YELLOW day
                     // TODO: go through the summary and check if the day is the first RED/YELLOW day
                 } else {
-                    //TODO: we are green again
-                    //TODO finish this decision branch
+                    when (warningType) {
+                        WarningType.RED -> quarantineRepository.revokeLastRedContactDate()
+                        WarningType.YELLOW -> quarantineRepository.revokeLastYellowContactDate()
+                    }
                 }
             }
             WarningType.REVOKE -> {
                 //we are above risc for the last days!!!
-                if (summary.summationRiskScore >= configuration.dailyRiskThreshold){
+                if (summary.summationRiskScore >= configuration.dailyRiskThreshold) {
                     //we must now identify day by day if we are YELLOW or RED
                     val listOfDaysWithDownloadedFilesSortedByServer = apiInteractor.fetchDailyBatchDiagnosisKeys()
+                    //process the batches and do one of these:
+                    //TODO find time of contact
+                    //this is a fake calculation:
+                    val dayOfExposure = ZonedDateTime.now().minusDays(summary.daysSinceLastExposure.toLong())
+                    quarantineRepository.receivedWarning(WarningType.RED, timeOfContact = dayOfExposure)
```

Comment:
`startOfTheDay()`?

Reply:
@dusanjencik

Reply:
@falkorichter please remember this objection